plugins/microsoft-graph-authz/src/main/java/org/elasticsearch/xpack/security/authz/microsoft/MicrosoftGraphAuthzRealm.java

```diff
@@ -43,8 +121,97 @@ public void authenticate(AuthenticationToken token, ActionListener<Authenticatio
+                        userProperties.v1(),
+                        userProperties.v2(),
+                        Map.of(),
+                        true
+                    );
+                    logger.trace("Authorized user from Microsoft Graph {}", user);
+                    l.onResponse(user);
+                }));
+            } catch (ReflectiveOperationException | ODataError e) {
+                logger.error("failed to authenticate with realm", e);
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private GraphServiceClient buildClient(SecureString clientSecret) {
+        final var credentialProviderBuilder = new ClientSecretCredentialBuilder().clientId(
+            config.getSetting(MicrosoftGraphAuthzRealmSettings.CLIENT_ID)
+        )
+            .clientSecret(clientSecret.toString())
+            .tenantId(config.getSetting(MicrosoftGraphAuthzRealmSettings.TENANT_ID))
+            .authorityHost(config.getSetting(MicrosoftGraphAuthzRealmSettings.ACCESS_TOKEN_HOST));
+
+        if (DISABLE_INSTANCE_DISCOVERY) {
+            credentialProviderBuilder.disableInstanceDiscovery();
+        }
+        final var credentialProvider = credentialProviderBuilder.build();
+
+        return new GraphServiceClient(
+            new BaseGraphRequestAdapter(
+                new AzureIdentityAuthenticationProvider(credentialProvider, Strings.EMPTY_ARRAY, "https://graph.microsoft.com/.default"),
+                config.getSetting(MicrosoftGraphAuthzRealmSettings.API_HOST)
+            )
+        );
+    }
+
+    private Tuple<String, String> fetchUserProperties(GraphServiceClient client, String userId) {
+        var response = client.users()
+            .byUserId(userId)
+            .get(requestConfig -> requestConfig.queryParameters.select = new String[] { "displayName", "mail" });
+
+        logger.trace("Fetched user with name [{}] and email [{}] from Microsoft Graph", response.getDisplayName(), response.getMail());
+
+        return Tuple.tuple(response.getDisplayName(), response.getMail());
+    }
+
+    private List<String> fetchGroupMembership(GraphServiceClient client, String userId) throws ReflectiveOperationException {
+        List<String> groups = new ArrayList<>();
+
+        var groupMembership = client.users().byUserId(userId).transitiveMemberOf().graphGroup().get(requestConfig -> {
+            requestConfig.queryParameters.select = new String[] { "id" };
+            requestConfig.queryParameters.top = 999;
+        });
+
+        var pageIterator = new PageIterator.Builder<Group, GroupCollectionResponse>().client(client)
+            .collectionPage(groupMembership)
+            .collectionPageFactory(GroupCollectionResponse::createFromDiscriminatorValue)
+            .requestConfigurator(requestInfo -> {
+                requestInfo.addQueryParameter("%24select", new String[] { "id" });
+                requestInfo.addQueryParameter("%24top", "999");
```

Comment:
Since this is used in two places and a little confusing a comment + constant would help.